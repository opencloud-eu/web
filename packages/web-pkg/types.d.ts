import { Ability } from '@opencloud-eu/web-client'
import {
  ArchiverService,
  ClientService,
  LoadingService,
  PreviewService,
  PasswordPolicyService
} from './src/services'
import { Route, Router } from 'vue-router'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
    $passwordPolicyService: PasswordPolicyService
    $router: Router
    $route: Route
  }
}
