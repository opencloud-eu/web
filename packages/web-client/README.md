# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
