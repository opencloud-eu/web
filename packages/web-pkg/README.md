# web-pkg

`web-pkg` is a package that provides utilities, most importantly a variety of components and composables, that can be useful when developing apps and extensions for OpenCloud Web.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-pkg --save-dev

$ pnpm add -D @opencloud-eu/web-pkg

$ yarn add @opencloud-eu/web-pkg --dev
```

It's recommended to install this package as a dev dependency because it's only really needed for providing autocompletion in your IDE and unit tests. In a runtime context, the OpenCloud Web runtime provides the actual implementation.
