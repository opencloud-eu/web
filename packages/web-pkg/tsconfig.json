{
  "extends": "@opencloud-eu/tsconfig",

  // necessary for building web-pkg standalone, otherwise we get errors with inferred types
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@vue/shared": ["node_modules/@vue/shared/"],
      "@codemirror/autocomplete": ["./node_modules/@codemirror/autocomplete"],
      "vue/jsx-runtime": ["./node_modules/vue/jsx-runtime"],
      "@popperjs/core": ["./../design-system/node_modules/@popperjs/core"]
    }
  },

  // include component types of the design-system
  "include": ["../design-system/types.d.ts"]
}
