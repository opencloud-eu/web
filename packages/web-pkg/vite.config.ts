import { join, resolve } from 'path'
import { defineConfig, searchForWorkspaceRoot } from 'vite'
import dts from 'vite-plugin-dts'
import { nodePolyfills } from 'vite-plugin-node-polyfills'
import vue from '@vitejs/plugin-vue'
import pkg from './package.json'

const projectRootDir = searchForWorkspaceRoot(process.cwd())
const external = [...Object.keys(pkg.dependencies)]

export default defineConfig({
  resolve: {
    alias: {
      crypto: join(projectRootDir, 'polyfills/crypto.js')
    }
  },
  css: {
    preprocessorOptions: {
      scss: {
        additionalData: `
          @use "sass:math";
          @use "sass:string";
          @use "sass:meta";
        `,
        silenceDeprecations: ['legacy-js-api', 'import']
      }
    }
  },
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-pkg',
      fileName: 'web-pkg'
    },
    rollupOptions: {
      external: external.filter(
        (e) =>
          // something is off with this lib, see https://github.com/ahmadjoya/generate-password-lite/issues/8
          e !== 'js-generate-password'
      )
    }
  },
  plugins: [
    vue(),
    nodePolyfills({
      exclude: ['crypto']
    }),
    dts({ exclude: ['**/tests'] })
  ]
})
