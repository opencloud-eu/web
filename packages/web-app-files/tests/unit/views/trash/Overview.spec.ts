import TrashOverview from '../../../../src/views/trash/Overview.vue'
import {
  defaultComponentMocks,
  defaultPlugins,
  defaultStubs,
  mount,
  RouteLocation
} from '@opencloud-eu/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { nextTick } from 'vue'
import { SpaceResource } from '@opencloud-eu/web-client'
import { ResourceTable, SortDir } from '@opencloud-eu/web-pkg'

const spaceMocks = [
  {
    id: '1',
    name: 'Personal',
    disabled: false,
    driveType: 'personal',
    getDriveAliasAndItem: () => '1',
    isOwner: () => true,
    hasTrashedItems: true
  },
  {
    id: '2',
    name: 'Project space 1',
    disabled: false,
    driveType: 'project',
    getDriveAliasAndItem: () => '2',
    isOwner: () => false,
    hasTrashedItems: true
  },
  {
    id: '3',
    name: 'Project space 2',
    disabled: false,
    driveType: 'project',
    getDriveAliasAndItem: () => '3',
    isOwner: () => false,
    hasTrashedItems: true
  }
] as unknown as SpaceResource[]

describe('TrashOverview', () => {
  it('should render no content message if no spaces exist', async () => {
    const { wrapper } = getWrapper({ spaces: [] })
    await (wrapper.vm as any).loadResourcesTask.last
    expect(wrapper.find('no-content-message-stub').exists()).toBeTruthy()
  })
  it('should navigate to single space trash if only one space exists', () => {
    const { mocks } = getWrapper({ spaces: [spaceMocks[0]] })
    expect(mocks.$router.push).toHaveBeenCalledWith({
      name: 'files-trash-generic',
      params: { driveAliasAndItem: spaceMocks[0].getDriveAliasAndItem(undefined) },
      query: {}
    })
  })
  describe('view states', () => {
    it('shows the loading spinner during loading', async () => {
      const { wrapper } = getWrapper()
      await nextTick()
      expect(wrapper.find('oc-spinner-stub').exists()).toBeTruthy()
    })
    it('should render trash list', async () => {
      const { wrapper } = getWrapper()
      await (wrapper.vm as any).loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('sorting', () => {
    it('sorts by property name', async () => {
      const { wrapper } = getWrapper()
      await (wrapper.vm as any).loadResourcesTask.last
      let sortedSpaces = []
      ;(wrapper.vm as any).sortBy = 'name'

      await nextTick()
      sortedSpaces = wrapper.findComponent(ResourceTable).props().resources as SpaceResource[]
      expect(sortedSpaces.map((s) => s.id)).toEqual([
        spaceMocks[0].id,
        spaceMocks[1].id,
        spaceMocks[2].id
      ])
      ;(wrapper.vm as any).sortDir = SortDir.Desc
      await nextTick()
      sortedSpaces = wrapper.findComponent(ResourceTable).props().resources as SpaceResource[]
      expect(sortedSpaces.map((s) => s.id)).toEqual([
        spaceMocks[0].id,
        spaceMocks[2].id,
        spaceMocks[1].id
      ])
    })
    it('should set the sort parameters accordingly when calling "handleSort"', () => {
      const { wrapper } = getWrapper({ spaces: [spaceMocks[0]] })
      const sortBy = 'name'
      const sortDir = SortDir.Desc
      ;(wrapper.vm as any).handleSort({ sortBy, sortDir })
      expect((wrapper.vm as any).sortBy).toEqual(sortBy)
      expect((wrapper.vm as any).sortDir).toEqual(sortDir)
    })
  })
  describe('filtering', () => {
    it('shows only filtered spaces if filter applied', async () => {
      const { wrapper } = getWrapper()
      ;(wrapper.vm as any).filterTerm = 'Personal'
      await nextTick()
      expect((wrapper.vm as any).displaySpaces.length).toEqual(1)
      expect((wrapper.vm as any).displaySpaces[0].id).toEqual(spaceMocks[0].id)
    })
  })
})

function getWrapper({ spaces = spaceMocks }: { spaces?: SpaceResource[] } = {}) {
  const mocks = {
    ...defaultComponentMocks({
      currentRoute: mock<RouteLocation>({ name: 'trash-overview' })
    })
  }

  mocks.$clientService.graphAuthenticated.drives.listMyDrives.mockResolvedValue(spaceMocks)

  return {
    mocks,
    wrapper: mount(TrashOverview, {
      global: {
        stubs: { ...defaultStubs, NoContentMessage: true, 'resource-table': false },
        mocks,
        provide: mocks,
        plugins: [...defaultPlugins({ piniaOptions: { spacesState: { spaces } } })]
      }
    })
  }
}
