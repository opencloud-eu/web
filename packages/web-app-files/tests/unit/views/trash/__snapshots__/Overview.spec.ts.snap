// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrashOverview > view states > should render trash list 1`] = `
"<div class="oc-flex">
  <div data-v-39c970be="" class="files-view-wrapper oc-width-expand">
    <div data-v-39c970be="" id="files-view">
      <app-bar-stub viewmodedefault="resource-table" breadcrumbs="[object Object]" breadcrumbscontextactionsitems="" viewmodes="" hasbulkactions="false" hasviewoptions="false" hashiddenfiles="false" hasfileextensions="false" haspagination="false" showactionsonselection="false" issidebaropen="false" batchactionsloading="false"></app-bar-stub>
      <div class="trash-bin-filters oc-flex oc-flex-right oc-flex-wrap oc-flex-bottom oc-mx-m oc-mb-m">
        <div class=""><label class="oc-label" for="spaces-filter">Search
            <!--v-if-->
          </label>
          <div class="oc-position-relative">
            <!--v-if--> <input id="spaces-filter" autocomplete="off" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text" value="">
            <!--v-if-->
          </div>
          <!--v-if-->
          <!--v-if-->
        </div>
      </div>
      <table class="oc-table oc-table-hover oc-table-sticky has-item-context-menu trash-table files-table" id="files-space-table" selection="">
        <thead class="oc-thead">
          <tr class="oc-table-header-row">
            <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-expand oc-th oc-table-header-cell oc-table-header-cell-name oc-pl-m" style="top: 0px;" aria-sort="ascending"><button type="button" aria-label="Sort by name" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-s oc-button-secondary oc-button-raw oc-button-secondary-raw no-hover oc-button-sort oc-width-1-1">
                <!--v-if-->
                <!-- @slot Content of the button --> <span class="oc-table-thead-content header-text">Name</span> <span class="oc-icon oc-icon-s oc-p-xs oc-rounded"><!----></span>
              </button></th>
            <th class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-shrink oc-th oc-table-header-cell oc-table-header-cell-indicators oc-pr-m" style="top: 0px;">
              <div><span class="oc-table-thead-content header-text">Status</span></div>
            </th>
          </tr>
        </thead>
        <tbody class="has-item-context-menu">
          <tr class="oc-tbody-tr oc-tbody-tr-1" data-item-id="1" draggable="false">
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td" colspan="2"><span class="shimmer"></span></td>
          </tr>
          <tr class="oc-tbody-tr oc-tbody-tr-2" data-item-id="2" draggable="false">
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td" colspan="2"><span class="shimmer"></span></td>
          </tr>
          <tr class="oc-tbody-tr oc-tbody-tr-3" data-item-id="3" draggable="false">
            <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td" colspan="2"><span class="shimmer"></span></td>
          </tr>
        </tbody>
        <tfoot class="oc-table-footer">
          <tr class="oc-table-footer-row">
            <td colspan="2" class="oc-table-footer-cell">
              <!-- @slot Footer of the table -->
              <!-- @slot Footer of the files table -->
              <div class="oc-text-center oc-width-1-1 oc-my-s">
                <p class="oc-text-muted">3 trashes in total</p>
                <!--v-if-->
              </div>
            </td>
          </tr>
        </tfoot>
      </table>
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
  <file-side-bar-stub isopen="false"></file-side-bar-stub>
</div>"
`;
